src/python/grpcio/grpc/_channel.py

```diff
@@ -0,0 +1,683 @@
+      if serialized_response is not None:
+        response = _common.deserialize(
+            serialized_response, response_deserializer)
+        if response is None:
+          state.code = grpc.StatusCode.INTERNAL
+          state.details = b'Exception deserializing response!'
+        else:
+          state.response = response
+    elif operation_type is cygrpc.OperationType.receive_status_on_client:
+      state.trailing_metadata = batch_operation.received_metadata
+      if state.code is None:
+        code = _common.CYGRPC_STATUS_CODE_TO_STATUS_CODE.get(
+            batch_operation.received_status_code)
+        if code is None:
+          state.code = grpc.StatusCode.UNKNOWN
+          state.details = _unknown_code_details(
+              batch_operation.received_status_code,
+              batch_operation.received_status_details)
+        else:
+          state.code = code
+          state.details = batch_operation.received_status_details
+      for callback in state.callbacks:
+        callback()
+
+
+def _event_handler(state, call, response_deserializer):
+  def handle_event(event):
+    with state.condition:
+      _handle_event(event, state, response_deserializer)
+      state.condition.notify_all()
+      if state.due:
+        return None
+      else:
+        return call
+  return handle_event
+
+
+def _drain(state, completion_queue, response_deserializer):
+  while state.due:
+    event = completion_queue.poll()
+    _handle_event(event, state, response_deserializer)
+
+
+def _consume_request_iterator(
+    request_iterator, state, call, request_serializer):
+  event_handler = _event_handler(state, call, None)
+  def consume_request_iterator():
+    for request in request_iterator:
+      serialized_request = _common.serialize(request, request_serializer)
+      with state.condition:
+        if state.code is None:
+          if serialized_request is None:
+            call.cancel()
+            state.code = grpc.StatusCode.INTERNAL
+            state.details = b'Exception serializing request!'
+            return
+          else:
+            operations = (cygrpc.operation_send_message(serialized_request),)
+            call.start_batch(cygrpc.Operations(operations), event_handler)
+            state.due.add(cygrpc.OperationType.send_message)
+            while True:
+              state.condition.wait()
+              if state.code is None:
+                if cygrpc.OperationType.send_message not in state.due:
+                  break
+              else:
+                return
+    with state.condition:
+      if state.code is None:
+        operations = (cygrpc.operation_send_close_from_client(),)
+        call.start_batch(cygrpc.Operations(operations), event_handler)
+        state.due.add(cygrpc.OperationType.send_close_from_client)
+  thread = threading.Thread(target=consume_request_iterator)
+  thread.start()
+
+
+class _Rendezvous(grpc.Future, grpc.Call, grpc.RpcError):
```

Comment:
I still find this whole inheritance-instead-of-composition thing enormously confusing, but without statically typed attributes my personal reservations are kinda moot...